### ISSUE TITLE
Types in Union are constructed incompatibly

### ISSUE THREAD

**Issue by jogibear9988** [linq2db/linq2db#4225]
### Describe your issue

I've the following code:

                var qry1 = from a in s.GetTable<UserDTO>()
                           select new { u = a, g = (GroupDTO)null };
                var qry2 = from a in s.GetTable<GroupDTO>()
                           select new { u = (UserDTO)null, g = a };
                var q = qry1.UnionAll(qry2).ToList();

I don't know, but should this work?

```
Exception message: Types in Union are constructed incompatibly
```

### Environment details

`Linq To DB` version: current

**Comment by HypeillingerChillinger** [comment]
What I do is

```csharp
var unionDoc = from a in myQuery
from b in context.B.Where(x => x.Code == a.Str)
from c in context.C.Where(x => false).Take(1).DefaultIfEmpty()
select new Bla
{
   B = b,
   C = c
};

var unionMac = from a in myQuery
from b in context.B.Where(x => false).Take(1).DefaultIfEmpty()
from c in context.C.Where(x => x.Code == a.Str)
select new Bla
{
   B = b,
   C = c
};

unionDoc.UnionAll(unionMac);
```

But it is really annoying to write *dummy* lines like this: `from b in context.B.Where(x => false).Take(1).DefaultIfEmpty()` in order to prevent the exception `Exception message: Types in Union are constructed incompatibly`.

What I intended to do is to prevent duplicates that are created in a normal join when joining from a root table to many other tables like

Table A holds 1 row
Table B holds 3 rows
Table C holds 3 rows

The result of a join is 9 rows.

With a union all I can reduce it to 7 rows.

Not sure what is faster. May be this is an option too: 
https://stackoverflow.com/questions/10916934/full-outer-join-using-each-row-once

**Comment by MaceWindu** [comment]
Fixed in `6.0.0-preview.1`
